### PR TITLE
CNS-574 add slot filters

### PR DESCRIPTION
**Pull request by omerlavanet**

### CODE REVIEW

File: x/pairing/keeper/filters/filter.go [hunk 1]
```diff
@@ -37,36 +39,99 @@ func initFilters(filters []Filter, strictestPolicy planstypes.Policy) []Filter {
...
 		for i := 0; i < len(filters); i++ {
-			result = result && filtersResult[i][j]
-			if !result {
-				break
+			// if filter result was negative meaning provider didn't pass
+			if !filtersResult[i][j] {
+				// check if filter is mandatory
+				if filters[i].IsMix() {
+					// filter is a mix filter, that didn't pass
+					for _, index := range mixFilterIndexes[filters[i]] {
```
Comment (Yaroms) [inline x/pairing/keeper/filters/filter.go:75]: rename to slotIndex for clarity

File: x/pairing/keeper/scores/pairing_score.go [hunk 2]
```diff
@@ -16,6 +16,28 @@ type PairingScore struct {
...
 	ScoreComponents  map[string]math.Uint
 	SkipForSelection bool
+	SlotFiltering    map[int]struct{} // slot indexes here are skipped
+}
+
+func (ps *PairingScore) IsValidForSelection(slotIndex int) bool {
+	if ps.SkipForSelection {
+		return false
+	}
+	if _, ok := ps.SlotFiltering[slotIndex]; ok {
```
Comment (Yaroms) [inline x/pairing/keeper/scores/pairing_score.go:26]: _, ok := ps.SlotFiltering[slotIndex]
return !ok

File: x/pairing/keeper/scores/pairing_slot.go [hunk 3]
```diff
@@ -2,20 +2,37 @@ package scores
...
+}
+
+func NewPairingSlotGroup(pairingSlot *PairingSlot) *PairingSlotGroup {
+	return &PairingSlotGroup{indexes: []int{pairingSlot.Index}, Reqs: pairingSlot.Reqs}
 }
 
-func NewPairingSlot() *PairingSlot {
-	return &PairingSlot{Count: 1}
+func (psg *PairingSlotGroup) AddToGroup(pairingSlot *PairingSlot) {
+	psg.indexes = append(psg.indexes, pairingSlot.Index)
```
Comment (Yaroms) [inline x/pairing/keeper/scores/pairing_slot.go:28]: verify that they have the same requirements?

File: scripts/init_chain_commands.sh [hunk 4]
```diff
@@ -12,14 +12,15 @@ GASPRICE="0.000000001ulava"
...
 lavad tx gov vote $(latest_vote) yes -y --from alice --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE &
-wait_next_block
 echo "voted on first proposal"
-
+sleep 4
 lavad tx gov submit-legacy-proposal plans-add ./cookbook/plans/default.json,./cookbook/plans/temporary-add.json -y --from alice --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE
 wait_count_blocks 2
 lavad tx gov vote $(latest_vote) yes -y --from alice --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE
 wait_next_block
+lavad tx gov vote 2 yes -y --from alice --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE
```
Comment (ranlavanet) [inline scripts/init_chain_commands.sh:23]: you are voting twice

File: scripts/init_chain_commands.sh [hunk 5]
```diff
@@ -12,14 +12,15 @@ GASPRICE="0.000000001ulava"
...
 echo "submitted first proposal"
+echo "latest vote2: $(latest_vote)"
 lavad tx gov vote $(latest_vote) yes -y --from alice --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE &
-wait_next_block
 echo "voted on first proposal"
-
+sleep 4
 lavad tx gov submit-legacy-proposal plans-add ./cookbook/plans/default.json,./cookbook/plans/temporary-add.json -y --from alice --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE
 wait_count_blocks 2
 lavad tx gov vote $(latest_vote) yes -y --from alice --gas-adjustment "1.5" --gas "auto" --gas-prices $GASPRICE
```
Comment (ranlavanet) [inline scripts/init_chain_commands.sh:21]: here 1

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: x/pairing/keeper/scores/pairing_slot.go (hunk 3 above)
Comment (omerlavanet) [reply x/pairing/keeper/scores/pairing_slot.go:28]: the only use of this is after checking equal no reason to check twice

**Review by oren-lava — DISMISSED** [review]

**Review by Yaroms — APPROVED** [review]

File: scripts/init_install.sh [hunk 6]
```diff
@@ -67,6 +67,15 @@ else
...
+if ! command_exists protoc-gen-gocosmos; then
+    git clone https://github.com/cosmos/gogoproto.git
+    cd gogoproto
+    go mod download
+    make install
+    cd ..
+    rm -rf gogoproto
+fi
+
 if ! command_exists yq; then
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply scripts/init_install.sh:79]: we have a different task for that

File: scripts/init_chain_commands.sh (hunk 4 above)
Comment (Yaroms) [reply scripts/init_chain_commands.sh]: done

File: cookbook/plans/free.json [hunk 7]
```diff
@@ -18,7 +18,9 @@
                     "geolocation_profile": "4",
                     "total_cu_limit": "10000000",
                     "epoch_cu_limit": "100000",
-                    "max_providers_to_pair": "5"
+                    "max_providers_to_pair": "5",
+                    "chain_policies": [
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply cookbook/plans/free.json]: lets do another task for that

**Pull request merged**